spinnaker/api/pipeline.go

```diff
@@ -8,6 +8,117 @@ import (
 	gate "github.com/spinnaker/spin/cmd/gateclient"
 )
 
+type PipelineConfig struct {
+	Pipeline
+	ID                string                   `json:"id,omitempty"`
+	Type              string                   `json:"type,omitempty"`
+	Name              string                   `json:"name"`
+	Application       string                   `json:"application"`
+	Triggers          []map[string]interface{} `json:"triggers,omitempty"`
+	ExpectedArtifacts []map[string]interface{} `json:"expectedArtifacts,omitempty"`
+	Notifications     []map[string]interface{} `json:"notifications,omitempty"`
+	LastModifiedBy    string                   `json:"lastModifiedBy"`
+	Config            interface{}              `json:"config,omitempty"`
+	UpdateTs          string                   `json:"updateTs,omitempty"`
+}
+
+type PipelineDocument struct {
+	Pipeline
+	AppConfig map[string]string `json:"appConfig,omitempty" mapstructure:"config"`
+}
+
+type PipelineParameter struct {
+	Description string   `json:"description,omitempty"`
+	Default     string   `json:"default,omitempty"`
+	Name        string   `json:"name"`
+	Required    bool     `json:"required"`
+	HasOptions  bool     `json:"hasOptions"`
+	Label       string   `json:"label,omitempty"`
+	Options     []string `json:"options,omitempty"`
+}
+
+type Pipeline struct {
+	Description          string               `json:"description,omitempty"`
+	ExecutionEngine      string               `json:"executionEngine,omitempty" mapstructure:"engine"`
+	Parallel             *bool                `json:"parallel,omitempty"`
+	LimitConcurrent      *bool                `json:"limitConcurrent,omitempty" mapstructure:"limit_concurrent"`
+	KeepWaitingPipelines *bool                `json:"keepWaitingPipelines,omitempty" mapstructure:"wait"`
+	Stages               []*Stage             `json:"stages,omitempty" mapstructure:"stage"`
+	Parameters           []*PipelineParameter `json:"parameterConfig,omitempty" mapstructure:"parameter"`
+}
+
+type Stage struct {
```

Comment:
I love where this is going but I'd like to see this abstracted a little bit more. For example, most of these fields can apply to _any_ stage (`FailPipeline`, `Instructions`, etc) but other fields only apply to what looks like a Deploy stage for the Kubernetes V1 provider. Could you define some concrete stage definitions for the stages you'd like to use instead of throwing them into a single `Stage` type?

Also, could you include an example of how this would be used?

Reply:
I've finally found some time to make changes as requested. Also added some more fields. I would consider this data source as "work in progress" since I expect to add more fields with time to match 1:1 to what spinnaker offers.